File: number-of-provinces/number-of-provinces.java
class Solution {
    int provinces = 0;
    boolean[] visited;
    public int findCircleNum(int[][] isConnected) {
        if (isConnected.length == 1) return 1;
        visited = new boolean[isConnected.length];
        for (int i = 0; i < isConnected.length; i++) {
            if (!visited[i]) {
                provinces++;
                dfs(i, isConnected);
            }
        }
        return provinces;
    }

    private void dfs(int node, int[][] isConnected) {
        visited[node] = true;
        for (int i = 0; i < isConnected[node].length; i++) {
            if (isConnected[node][i] == 1 && node != i && !visited[i]) {
                dfs(i, isConnected);
            }
        }
    }
}

File: 547-number-of-provinces/number-of-provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        provinces = 0
        visited = [0] * len(isConnected)
        def explore(i):
            visited[i] = 1
            for j in range(len(isConnected)):
                if isConnected[i][j] == 1 and not visited[j]:
                    explore(j)
        for i in range(len(isConnected)):
            if not visited[i]:
                provinces += 1
                explore(i)
        return provinces